Enhancement: implement document link custom field

shamoon opened the pull request:
<!--
Note: All PRs with code changes should be targeted to the `dev` branch, pure documentation changes can target `main`
-->

## Proposed change

<img width="765" alt="Screenshot 2023-12-03 at 11 03 13 PM" src="https://github.com/paperless-ngx/paperless-ngx/assets/4887959/826988f7-8b73-41bd-9562-c8a27282e0a6">
<img width="510" alt="Screenshot 2023-12-04 at 7 59 31 PM" src="https://github.com/paperless-ngx/paperless-ngx/assets/4887959/0a982757-f179-4fe4-b3af-b9cc9bd3074e">

This was a fun little thing to do, most of it is frontend and made easier by one of our lovelier dependencies, `ng-select`, mostly seemed like hooking existing things up. Some notes:

- ~~Biggest debate I had was what type of Django field to use, I didnt want to deal with ManyToMany drama etc. so ultimately I decided on a list of the doc IDs, closest way to store that is a char field with the validator, I then parse it with JSON to make the API output cleaner.~~ Edit: JSONField. Also I chose to re-create the migration 1042 since its unreleased
- When loading the value we have to hit the API for the documents (e.g. title) async, as well as searching
- Can search by title, I thought search by ID (like GH) while cool, seemed unnecessary

Welcome any feedback.

Fixes #422

## Type of change

<!--
What type of change does your PR introduce to Paperless-ngx?
NOTE: Please check only one box!
-->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Other (please explain): Enhancement

## Checklist:

<!--
NOTE: PRs that do not address the following will not be merged, please do not skip any relevant items.
-->

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [x] If applicable, I have included testing coverage for new code in this PR, for [backend](https://docs.paperless-ngx.com/development/#testing) and / or [front-end](https://docs.paperless-ngx.com/development/#testing-and-code-style) changes.
- [x] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [x] If applicable, I have checked that all tests pass, see [documentation](https://docs.paperless-ngx.com/development/#back-end-development).
- [x] I have run all `pre-commit` hooks, see [documentation](https://docs.paperless-ngx.com/development/#code-formatting-with-pre-commit-hooks).
- [x] I have made corresponding changes to the documentation as needed.
- [x] I have checked my modifications for any breaking changes.

shamoon committed:
Feature: implement document link custom field

shamoon committed:
Update messages.xlf

stumpylog commented on src/documents/serialisers.py:
```diff
@@ -458,7 +460,11 @@ def create(self, validated_data):
+        return (
+            obj.value
+            if (obj.field.data_type != CustomField.FieldDataType.DOCUMENTLINK)
+            else json.loads(obj.value)
```
Since it looks like you are using the new field as a JSON field, why not use a `JSONField`?

shamoon commented:
> Any chance to add the correspondent and document creation date to the dropdown as well?

I think we need to be very sparing about what data we show in the dropdown as it’s not big. I’ll look at showing the date but correspondent seems arbitrary, other people will want doc type, etc.

shamoon committed:
Use JSON field for document ids

shamoon committed:
Include document created in filter list

shamoon commented:
Ok, I converted to JSON field, also added the created date to the filter list, which I think is useful

LMK if anything else!

<img width="510" alt="Screenshot 2023-12-04 at 7 59 31 PM" src="https://github.com/paperless-ngx/paperless-ngx/assets/4887959/0a982757-f179-4fe4-b3af-b9cc9bd3074e">

shamoon committed:
Fix remove button alignment

shamoon committed:
100% coverage

shamoon committed:
Update frontend strings

jord1e commented:
🙏 maybe call it “References”?

shamoon commented:
Its a custom field, you can give it any name you want.

stumpylog commented:
Codecov seems confused here, feel free to ignore

shamoon commented:
Ok yea I just noticed that, seems it’s just being weird 👍

jord1e commented:
Create a custom field in the left menu and then in the top right on the document page click on "custom fields" to add it

![grafik](https://github.com/paperless-ngx/paperless-ngx/assets/30464310/268fa18f-0fda-4d5a-8e54-994538198c72)
